mountpoint-s3/src/fs.rs

```diff
@@ -603,16 +613,29 @@ where
         ino: InodeNo,
         atime: Option<OffsetDateTime>,
         mtime: Option<OffsetDateTime>,
+        _size: Option<u64>,
         _flags: Option<u32>,
     ) -> Result<Attr, Error> {
         tracing::info!(
-            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?}",
+            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?} size {:?}",
             ino,
             _flags,
             atime,
-            mtime
+            mtime,
+            _size
         );
-        let lookup = self.superblock.setattr(&self.client, ino, atime, mtime).await?;
+        let setattr_result = self.superblock.setattr(&self.client, ino, atime, mtime).await;
+        let lookup = match (setattr_result, _size, self.config.allow_overwrite) {
+            (Ok(lookup), _, _) => lookup,
+            (Err(InodeError::SetAttrNotPermittedOnRemoteInode(_)), Some(0), false) => {
+                // We want to provide better feedback to users to prompt them to opt-in to file overwrites if it looks like what the application needs.
+                // Instead of complex logic to match `setattr` truncation only, we just check for the error and if the size was set in the request.
+                // If so, we assume its probably a truncation.
+                return Err(
+                    err!(libc::EPERM, "file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it"));
+            }
+            (Err(e), _, _) => return Err(e.into()),
+        };
```

Comment:
There's a cool thing with the match statement you can do which is add an if condition using a 'match guard'. https://doc.rust-lang.org/reference/expressions/match-expr.html#match-guards

Try it out for the `allow_overwrite` check, I think it'll make the match neater and easier to follow.